src/hdr_histogram.c

```diff
@@ -405,68 +428,140 @@ void hdr_init_preallocated(struct hdr_histogram* h, struct hdr_histogram_bucket_
     h->total_count                     = 0;
 }
 
+int yb_hdr_init(
+        int64_t lowest_discernible_value,
+        int64_t highest_trackable_value,
+        int yb_bucket_factor,
+        hdr_histogram* histogram)
+{
+    hdr_histogram_bucket_config cfg;
+
+    int r = yb_hdr_calculate_bucket_config(lowest_discernible_value, highest_trackable_value, yb_bucket_factor, &cfg);
+    if (r)
+    {
+        return r;
+    }
+
+    if (!histogram)
+    {
+        return ENOMEM;
+    }
+
+    hdr_init_preallocated(histogram, &cfg);
+
+    return 0;
+}
+
 int hdr_init(
         int64_t lowest_discernible_value,
         int64_t highest_trackable_value,
         int significant_figures,
-        struct hdr_histogram** result)
+        hdr_histogram** result)
 {
-    int64_t* counts;
-    struct hdr_histogram_bucket_config cfg;
-    struct hdr_histogram* histogram;
+
+    hdr_histogram_bucket_config cfg;
+    hdr_histogram* histogram;
 
     int r = hdr_calculate_bucket_config(lowest_discernible_value, highest_trackable_value, significant_figures, &cfg);
     if (r)
     {
         return r;
     }
 
-    counts = (int64_t*) hdr_calloc((size_t) cfg.counts_len, sizeof(int64_t));
+#ifdef FLEXIBLE_COUNTS_ARRAY
+    histogram = (hdr_histogram*) hdr_calloc(1, sizeof(hdr_histogram) + cfg.counts_len * sizeof(count_t));
```

Comment:
You probably need to do some compile time assertion on the `hdr_histogram` size to ensure the padding is as expected. This will avoid allocating counts array unaligned esp we want to persist hdr out to disk.

Reply:
Looked it up and could not find any readily available compile-time asserts in C. Let me know if you know otherwise. Added runtime asserts to yb_hdr_init and hdr_init instead.

Reply:
You can use `static_assert` here. It does compile time checking.

Reply:
Unfortunately static_assert doesn't exist in C99. Looks like it was introduced in C11: https://learn.microsoft.com/en-us/cpp/c-language/static-assert-c?view=msvc-170